templates/ikfast61_moveit_plugin_template.cpp

```diff
@@ -958,6 +1044,202 @@ bool IKFastKinematicsPlugin::getPositionIK(const geometry_msgs::Pose &ik_pose,
+    numsol = solve(frame,vfree,ik_solutions);
+    solution_set.push_back(ik_solutions);
+  }
+
+  ROS_DEBUG_STREAM_NAMED("ikfast","Found " << numsol << " solutions from IKFast");
+  bool solutions_found = false;
+  std::stringstream ss;
+  if( numsol > 0 )
+  {
+    for(unsigned int r = 0; r < solution_set.size() ; r++)
+    {
+
+      ik_solutions = solution_set[r];
+      numsol = ik_solutions.GetNumSolutions();
+      for(int s = 0; s < numsol; ++s)
+      {
+        std::vector<double> sol;
+        getSolution(ik_solutions,s,sol);
+        ss.str("");
+        ss<<"[";
+        for(unsigned int i = 0 ; i < sol.size() ; i++)
+        {
+          ss<<sol[i]<<" ";
+        }
+        ss<<"]";
+        ROS_DEBUG_NAMED("ikfast","Sol %d: %s", s, ss.str().c_str());
+
+        bool obeys_limits = true;
+        for(unsigned int i = 0; i < sol.size(); i++)
+        {
+          // Add tolerance to limit check
+          if(joint_has_limits_vector_[i] && ( (sol[i] < (joint_min_vector_[i]-LIMIT_TOLERANCE)) ||
+                                              (sol[i] > (joint_max_vector_[i]+LIMIT_TOLERANCE)) ) )
+          {
+            // One element of solution is not within limits
+            obeys_limits = false;
+            ROS_DEBUG_STREAM_NAMED("ikfast","Not in limits! " << i << " value " << sol[i] << " has limit: " << joint_has_limits_vector_[i] << "  being  " << joint_min_vector_[i] << " to " << joint_max_vector_[i]);
+            break;
+          }
+        }
+        if(obeys_limits)
+        {
+          // All elements of solution obey limits
+          solutions_found = true;
+          solutions.push_back(sol);
+        }
+      }
+    }
+
+    if(solutions_found)
+    {
+      result.kinematic_error = kinematics::KinematicErrors::OK;
+      return true;
+    }
+  }
+  else
+  {
+    ROS_DEBUG_STREAM_NAMED("ikfast","No IK solution");
+  }
+
+  result.kinematic_error = kinematics::KinematicErrors::NO_SOLUTION;
+  return false;
+}
+
+bool IKFastKinematicsPlugin::sampleRedundantJoint(kinematics::DiscretizationMethod method, std::vector<double>& sampled_joint_vals) const
+{
+  double joint_min = -M_PI;
+  double joint_max = M_PI;
+  int index =  redundant_joint_indices_.front();
+  double joint_dscrt = redundant_joint_discretization_.at(index);
+
+  if(joint_has_limits_vector_[redundant_joint_indices_.front()])
+  {
+    joint_min = joint_min_vector_[index];
+    joint_max = joint_max_vector_[index];
+  }
+
+
+  switch(method)
+  {
+    case kinematics::DiscretizationMethods::ALL_DISCRETIZED:
+    {
+      int steps = std::ceil((joint_max - joint_min)/joint_dscrt);
+      for(unsigned int i = 0; i < steps;i++)
+      {
+        sampled_joint_vals.push_back(joint_min + joint_dscrt*i);
+      }
+      sampled_joint_vals.push_back(joint_max);
+    }
+      break;
+    case kinematics::DiscretizationMethods::ALL_RANDOM_SAMPLED:
+    {
+      int count = joint_dscrt > 1 ? static_cast<int>(joint_dscrt) : 1;
```

Comment:
The latest commit addresses this by using the joint discretization to determine how many random samples to generate as it is done for the uniform discretization.